### PR TITLE
FUSETOOLS2-304 - provide minikube on Travis CI

### PR DESCRIPTION
**Pull request by apupier**
Signed-off-by: Aurélien Pupier <apupier@redhat.com>

### CODE REVIEW

**Comment by apupier** [comment]
just having a minikube instance and kubectl on command-lien seems to cause trouble for tests, seems strange

```
 7 failing
  1) Camel K Task Completion
       Completion for traits:
     Error: Apache Camel K CLI (kamel) unavailable
  	at /home/travis/build/camel-tooling/vscode-camelk/out/src/kamel.js:9:4951
  	at Generator.next (<anonymous>)
  	at fulfilled (out/src/kamel.js:9:988)
  2) Should do completion in tasks.json
       Completes for traits:
     Error: Timed out after waiting for 10000ms
  	at Timeout.onTimeout (node_modules/async-wait-until/src/waitUntil.js:53:14)
  	at listOnTimeout (internal/timers.js:531:17)
  	at processTimers (internal/timers.js:475:7)
  3) ensure install methods are functioning as expected
       ensure cli version checking works correctly:
      AssertionError [ERR_ASSERTION]: cli version checking, positive case worked
      + expected - actual
      -false
      +true
      
  	at /home/travis/build/camel-tooling/vscode-camelk/out/src/test/suite/install.test.js:73:24
  4) ensure install methods are functioning as expected
       ensure we have access to the kamel cli:
      AssertionError [ERR_ASSERTION]: false == true
      + expected - actual
      -false
      +true
      
  	at Context.<anonymous> (out/src/test/suite/install.test.js:80:16)
  	at processImmediate (internal/timers.js:439:21)
  5) ensure install methods are functioning as expected
       ensure we have access to the kubectl cli:
      AssertionError [ERR_ASSERTION]: false == true
      + expected - actual
      -false
      +true
      
  	at Context.<anonymous> (out/src/test/suite/install.test.js:86:16)
  	at processImmediate (internal/timers.js:439:21)
  6) IntegrationUtil tests
       "after each" hook in "IntegrationUtil tests":
     Uncaught spawn kamel ENOENT (and Mocha's done() called multiple times)
  Error: spawn kamel ENOENT
  	at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
  	at onErrorNT (internal/child_process.js:456:16)
  	at processTicksAndRejections (internal/process/task_queues.js:77:11)
  7) ensure kamel and kubectl are available
       ensure can activate kamel cli:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/travis/build/camel-tooling/vscode-camelk/out/src/test/suite/kamel.test.js)
  	at listOnTimeout (internal/timers.js:531:17)
  	at processTimers (internal/timers.js:475:7)
```

**Comment by apupier** [comment]
this error is repeated several times. it seems to happen on the first test that requires the extension to be activated:

```
      ✓ Completes for Camel K template: 1231ms
Apache Camel K CLI (kamel) unavailable: Error: Apache Camel K CLI (kamel) unavailable
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/kamel.js:9:4951
	at Generator.next (<anonymous>)
	at fulfilled (/home/travis/build/camel-tooling/vscode-camelk/out/src/kamel.js:9:988)
```

**Comment by apupier** [comment]
another stack:
```
  ensure kamel and kubectl are available
Apache Camel K CLI (kamel) unavailable: Error: Apache Camel K CLI (kamel) unavailable
rejected promise not handled within 1 second: TypeError: Cannot read property 'catch' of undefined
stack trace: TypeError: Cannot read property 'catch' of undefined
	at CamelKNodeProvider.<anonymous> (/home/travis/build/camel-tooling/vscode-camelk/out/src/CamelKNodeProvider.js:9:3995)
	at Generator.next (<anonymous>)
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/CamelKNodeProvider.js:9:1746
	at new Promise (<anonymous>)
	at __awaiter (/home/travis/build/camel-tooling/vscode-camelk/out/src/CamelKNodeProvider.js:9:662)
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/CamelKNodeProvider.js:9:3601
	at new Promise (<anonymous>)
	at CamelKNodeProvider.addChild (/home/travis/build/camel-tooling/vscode-camelk/out/src/CamelKNodeProvider.js:9:3524)
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/test/suite/integrationExplorer.test.js:46:33
```

**Comment by apupier** [comment]
this stack might means that the extension cannot activate successfully (and would explain several of the failures)

```
stack trace: TypeError: Cannot read property 'catch' of undefined
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/extension.js:9:24169
	at Generator.next (<anonymous>)
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/extension.js:9:1746
	at new Promise (<anonymous>)
	at __awaiter (/home/travis/build/camel-tooling/vscode-camelk/out/src/extension.js:9:662)
	at /home/travis/build/camel-tooling/vscode-camelk/out/src/extension.js:9:23506
	at l.fire (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:845)
	at p.setVisible (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:528:364)
	at t.ExtHostTreeViews.$setVisible (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:524:600)
	at p._doInvokeHandler (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:688:464)
	at p._invokeHandler (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:688:156)
	at p._receiveRequest (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:686:817)
	at p._receiveOneMessage (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:685:623)
	at /home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:683:791
	at l.fire (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:845)
	at v.fire (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:191:325)
	at /home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:851:104
	at l.fire (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:845)
	at v.fire (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:191:325)
	at t.PersistentProtocol._receiveMessage (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:195:717)
	at /home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:192:897
	at l.fire (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:845)
	at p.acceptChunk (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:188:897)
	at /home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:188:249
	at Socket.t (/home/travis/build/camel-tooling/vscode-camelk/.vscode-test/vscode-1.43.1/VSCode-linux-x64/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:197:232)
	at Socket.emit (events.js:203:13)
	at addChunk (_stream_readable.js:295:12)
	at readableAddChunk (_stream_readable.js:276:11)
	at Socket.Readable.push (_stream_readable.js:210:10)
	at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
```

**Comment by apupier** [comment]
most of the tests are also failing on master... so not related to this PR:

```
 6 failing
  1) Camel K Task Completion
       Completion for traits:
     Error: Apache Camel K CLI (kamel) unavailable
  	at /home/travis/build/camel-tooling/vscode-camelk/out/src/kamel.js:9:4951
  	at Generator.next (<anonymous>)
  	at fulfilled (out/src/kamel.js:9:988)
  2) Should do completion in tasks.json
       Completes for traits:
     Error: Timed out after waiting for 10000ms
  	at Timeout.onTimeout (node_modules/async-wait-until/src/waitUntil.js:53:14)
  	at listOnTimeout (internal/timers.js:531:17)
  	at processTimers (internal/timers.js:475:7)
  3) ensure install methods are functioning as expected
       ensure cli version checking works correctly:
     AssertionError [ERR_ASSERTION]: Failed to establish Apache Camel K stable version: Response code 403 (rate limit exceeded)
  	at Context.<anonymous> (out/src/test/suite/install.test.js:67:24)
  	at Generator.next (<anonymous>)
  	at fulfilled (out/src/test/suite/install.test.js:21:58)
  4) ensure install methods are functioning as expected
       ensure we have access to the kamel cli:
      AssertionError [ERR_ASSERTION]: false == true
      + expected - actual
      -false
      +true
      
  	at Context.<anonymous> (out/src/test/suite/install.test.js:80:16)
  	at processImmediate (internal/timers.js:439:21)
  5) IntegrationUtil tests
       ensure filtering out Camel K task with unrelated target file when accessing 'Start Apache Camel Integration':
     Uncaught Error: spawn kamel ENOENT
  	at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
  	at onErrorNT (internal/child_process.js:456:16)
  	at processTicksAndRejections (internal/process/task_queues.js:77:11)
  6) ensure kamel and kubectl are available
       ensure can activate kamel cli:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/travis/build/camel-tooling/vscode-camelk/out/src/test/suite/kamel.test.js)
  	at listOnTimeout (internal/timers.js:531:17)
  	at processTimers (internal/timers.js:475:7)
```

**Comment by apupier** [comment]
current state:

- tests are working on CI
- it requires:
  - no kamel on system path
  - no kubectl on system path
  - kubeconfig configured locally to reach an instance where Camel K is installed

The constraint are a bit hard and tests are failing with not easily understandable errors if they are not met. I will continue to at least provide a better error message. If possible have a more robust test working with different environment.

**Comment by apupier** [comment]
on Ci, it is now failing with:

```
1) Kubectl integration watcher
       Check there is one message logged in case of connection error:
     expected shareMessage to be called once but was called twice
    shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, W0408 12:33:20.838847   10419 loader.go:223] Config not found: /home/travis/.kube/config
) at Socket.<anonymous> (/home/travis/build/camel-tooling/vscode-camelk/out/src/kubectl.js:9:8389)
    shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, error: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
) at Socket.<anonymous> (/home/travis/build/camel-tooling/vscode-camelk/out/src/kubectl.js:9:8389)
  AssertError: expected shareMessage to be called once but was called twice
      shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, W0408 12:33:20.838847   10419 loader.go:223] Config not found: /home/travis/.kube/config
  ) at Socket.<anonymous> (out/src/kubectl.js:9:8389)
  
      shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, error: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
  ) at Socket.<anonymous> (out/src/kubectl.js:9:8389)
  	at Object.fail (node_modules/sinon/lib/sinon/assert.js:107:21)
  	at failAssertion (node_modules/sinon/lib/sinon/assert.js:66:16)
  	at Object.assert.<computed> [as calledOnce] (node_modules/sinon/lib/sinon/assert.js:92:13)
  	at Context.<anonymous> (out/src/test/suite/kubectlwatcher.test.js:70:26)
  	at Generator.next (<anonymous>)
  	at fulfilled (out/src/test/suite/kubectlwatcher.test.js:21:58)
  2) Kubectl integration watcher
       Check there is only one message logged in case of connection error with View visible:
     expected shareMessage to be called once but was called twice
    shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, W0408 12:33:23.394903   10587 loader.go:223] Config not found: /home/travis/.kube/config
) at Socket.<anonymous> (/home/travis/build/camel-tooling/vscode-camelk/out/src/kubectl.js:9:8389)
    shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, error: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
) at Socket.<anonymous> (/home/travis/build/camel-tooling/vscode-camelk/out/src/kubectl.js:9:8389)
  AssertError: expected shareMessage to be called once but was called twice
      shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, W0408 12:33:23.394903   10587 loader.go:223] Config not found: /home/travis/.kube/config
  ) at Socket.<anonymous> (out/src/kubectl.js:9:8389)
  
      shareMessage({ _channel: [Promise] {  }, name: "Apache Camel K" }, error: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
  ) at Socket.<anonymous> (out/src/kubectl.js:9:8389)
  	at Object.fail (node_modules/sinon/lib/sinon/assert.js:107:21)
  	at failAssertion (node_modules/sinon/lib/sinon/assert.js:66:16)
  	at Object.assert.<computed> [as calledOnce] (node_modules/sinon/lib/sinon/assert.js:92:13)
  	at Context.<anonymous> (out/src/test/suite/kubectlwatcher.test.js:90:26)
  	at fulfilled (out/src/test/suite/kubectlwatcher.test.js:21:58)
```

locally, it is working fine. Currently, I don't understand why the error message is logged two times instead of a single time.

File: .travis.yml [hunk 1]
```diff
@@ -12,6 +22,27 @@ install:
   - 'npm run vscode:prepublish'
   - 'npm install -g vsce'
   - 'vsce package'
+before_script:
+  # Download and provide in path kubectl , which is a requirement for using minikube.
+  - curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/v1.17.4/bin/linux/amd64/kubectl && chmod +x kubectl && sudo mv kubectl /usr/local/bin/
```
Comment (apupier) [inline .travis.yml:27]: note: staying with 1.77.x for now due to 1.18.x having more requirements: https://github.com/lilic/travis-minikube/issues/24

**Comment by apupier** [comment]
@lhein @bfitzpat please test locally that I have not missed some other cases

**Comment by apupier** [comment]
@lhein @bfitzpat can you review this PR please?

**Comment by bfitzpat** [comment]
> @lhein @bfitzpat can you review this PR please

So playing with it locally, I have some instability with the `Should do completion in Camel K standalone files` tests - specifically java. Bumped up the JAVA_EXTENSION_READINESS_TIMEOUT a few times and it's still unstable, but it does run sometimes. 

That said, when that test ran ok, the tests seemed to run ok against a local minikube in these states:

* minikube stopped
* minikube deleted
* minikube started

File: src/test/suite/completion.java.test.ts [hunk 2]
```diff
@@ -19,10 +19,15 @@
 import * as vscode from 'vscode';
 import { areJavaDependenciesDownloaded } from '../../JavaDependenciesManager';
 import { getDocUri, checkExpectedCompletion } from './completion.util';
+import { fail } from 'assert';
 
 const os = require('os');
 const waitUntil = require('async-wait-until');
 
+const DOWNLOAD_JAVA_DEPENDENCIES_TIMEOUT = 60000;
```
Comment (bfitzpat) [inline src/test/suite/completion.java.test.ts:27]: I had to bump this to 120000 for it to work more regularly on my system, but not sure where the instability is coming from

Comment (apupier) [reply src/test/suite/completion.java.test.ts:27]: it was already the case before this PR, I will increase it in this one as i already increased it in this PR

**Pull request merged**